### ISSUE TITLE
[Good First Issue]: Align behavior of ONNX Frontend operator BatchNormalization-6, 9, 14, 15 with original framework

### ISSUE THREAD

**Issue by gkrivor** [openvinotoolkit/openvino#20554]
### Context

Neural networks are graphs consisting of nodes called operators. Each operator corresponds to a mathematical function, usually described in framework's documentation or an AI standard, such as ONNX.
OpenVINO ONNX Frontend is a component responsible for working with ONNX graphs and requires implementation of different ONNX operators in order to use ONNX models.
This task requires alignment between OpenVINO ONNX Frontend and original framework implementations of BatchNormalization for next list of opsets: [opset 6](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-6),  [opset 9](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-9),  [opset 14](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-14),  [opset 15](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-15)
Necessary help will be provided by ONNX Fronted team.

### What needs to be done?

Operator details can be found in [ONNX Operators](https://github.com/onnx/onnx/blob/main/docs/Operators.md#BatchNormalization)
More details can be found in ONNX Changelog: [opset 6](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-6),  [opset 9](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-9),  [opset 14](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-14),  [opset 15](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#BatchNormalization-15)
1. Operator already has a common implementation in OpenVINO. First of all, you need to review a documentation and prepare a table with differences between versions. It could be, for instance, a missing property, extended/reduced coverage of existing property, etc...
2. Copy existing implementation [here](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op) to make it aligned with original framework (extend or reduce coverage of a common implementation). Copy of modified implementation should be in a defined opset, or in opset 1 in case it implements oldest implementation. [Example of multi-opset operation](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp#L50).
3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560) while keeping alphabetical order
4. Create test model(s) in [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/). OpenVINO test infrastructure then converts prototxt files to ONNX models - you will use those models later in tests
5. Add tests covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import.in.cpp)
6. Check [Python xfailed tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/tests_python/test_backend.py) to find a test marked as a xfailed for added functionality. If any exist - remove corresponding lines and try to verify by using cmdline "python -m pytest -k name_of_test".
More details in [adding operators to ONNX Frontend guide](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md)

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Intel DevHub Discord channel](https://discord.gg/wPuqAujS) - engage in discussions, ask questions and talk to OpenVINO developers
- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [User documentation](https://docs.openvino.ai/)
- Guides on [ONNX Frontend architecture](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/architecture.md) and [ONNX Frontend tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/tests.md) are also available
- You can also see a [blog post on contributing to OpenVINO](https://medium.com/openvino-toolkit/how-to-contribute-to-an-ai-open-source-project-c741f48e009e)

### Contact points

@gkrivor

**Comment by dev-seek** [comment]
.take

**Comment by dev-seek** [comment]
![Image](https://github.com/openvinotoolkit/openvino/assets/101525769/020ab128-0bda-48b1-91f2-239addd92d9f)

why i got this error and what actions can i take to resolve this error

**Comment by p-wysocki** [comment]
Hello @dev-seek, does this happen when running `git submodule update --init`?

A user [here](https://stackoverflow.com/questions/67793163/error-decoding-the-received-tls-packet-when-cloning-into-git-repository) says that it can be fixed by changing the method to SSH from HTTPS:
![image](https://github.com/openvinotoolkit/openvino/assets/55858107/77967a39-8d97-499a-87f6-39e31ced9ff5)
Overall this seems to me like some limitation of your network - another user say that it might be due to some complaining proxy or a timeout.

**Comment by dev-seek** [comment]
yep its occour when i am using  git submodule update --init  , and ssh too didnt work , what to do now ? is there any other way ?

**Comment by dev-seek** [comment]
Funally Issue resolved ✌

**Comment by p-wysocki** [comment]
Great! Could you share what was the solution?

**Comment by dev-seek** [comment]
Its just the issue of college internet , using my own data solves the problem😅

**Comment by dev-seek** [comment]
![Image](https://github.com/openvinotoolkit/openvino/assets/101525769/0e5a9f27-6cd1-464e-8410-b36cf85fedc7)

I didnt  get the point 2  of **what need to be done** completely , did it asking for making changes in the above pointed files or i am wrong somewhere ,
What it mean by copying existing implementation , its already there and if it asking of making new file then what the need of doing that ?, liittle bit confused , needed some explanation please

**Comment by p-wysocki** [comment]
Could you please take a look @gkrivor?

**Comment by dev-seek** [comment]
@p-wysocki  @gkrivor  sry for disturbance , but can u plz clear my doubts asap

**Comment by dev-seek** [comment]
"This task requires alignment between OpenVINO ONNX Frontend and original framework implementations of BatchNormalization"

how to know the original framework implementation , little bit confused please help .

Does it means that till now we didnt have the implimentation of BatchNormalization which align with original framework implimentations ?

And also frameworks here means (PYTORCH , TENSORFLOW) etc , isn't it ?

**Comment by dev-seek** [comment]
@p-wysocki can you please re-send the invite of that discord server as the old invite get expired .

**Comment by gkrivor** [comment]
@dev-seek hi! I'll add link to a issue with similar discussion: https://github.com/openvinotoolkit/openvino/issues/20553

Yes, you need to touch both of files. You should achieve something like this : https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp which will implement a found differences.

Also you will need to extend a tests for exact implementations like in this PR: https://github.com/openvinotoolkit/openvino/pull/21825

I mean you need to add a prototxt file and corresponding unit test to verify a correctness.

**Comment by p-wysocki** [comment]
Hi @dev-seek, are you still working on that issue?

**Comment by p-wysocki** [comment]
Hello @dev-seek, are there any updates?

**Comment by dev-seek** [comment]
Hey @p-wysocki really sry for the delay 
Make this issue open again... 
I wont geeting sufficient time for this as i am busy in somthing else

**Comment by Bepitic** [comment]
.take

**Comment by Bepitic** [comment]
Here is a List of changes for the versions:
| Feature / Version          | BatchNormalization-1                          | BatchNormalization-6                          | BatchNormalization-7                          | BatchNormalization-9                          | BatchNormalization-14                         | BatchNormalization-15                         |
|----------------------------|------------------------------------------------|------------------------------------------------|------------------------------------------------|------------------------------------------------|------------------------------------------------|------------------------------------------------|
| **Attributes**             | consumed_inputs, epsilon, is_test, momentum, spatial | epsilon, is_test, momentum, spatial           | epsilon, momentum, spatial                    | epsilon, momentum                             | epsilon, momentum, training_mode               | epsilon, momentum, training_mode               |
| **Outputs**                | Y, mean, var, saved_mean, saved_var (training); Y (test) | Y, mean, var, saved_mean, saved_var (training); Y (test) | Y, mean, var, saved_mean, saved_var (training); Y (test) | Y, mean, var, saved_mean, saved_var (training); Y (test) | Y, running_mean, running_var (training); Y (test) | Y, running_mean, running_var (training); Y (test) |
| **Inputs**                 | X, scale, B, mean, var                         | X, scale, B, mean, var                         | X, scale, B, mean, var                         | X, scale, B, mean, var                         | X, scale, B, input_fmean, input_var             | X, scale, B, input_mean, input_var             |
| **Type Constraints**       | **T**: tensor(float16), tensor(float), tensor(double) | **T**: tensor(float16), tensor(float), tensor(double) | **T**: tensor(float16), tensor(float), tensor(double) | **T**: tensor(float16), tensor(float), tensor(double) | **T**: tensor(float16), tensor(float), tensor(double), tensor(bfloat16); **U**: tensor(float16), tensor(float), tensor(double), tensor(bfloat16) | **T**: tensor(float16), tensor(float), tensor(double), tensor(bfloat16); **T1**: tensor(float16), tensor(float), tensor(double), tensor(bfloat16); **T2**: tensor(float16), tensor(float), tensor(double), tensor(bfloat16) |
||||||
| **Changes**          | Attrib: *consumed_inputs* -> list of ints, legacy purposes. | Attrib: Removed *consumed_inputs* | Attrib: Removed *is_test* | Attrib: Removed *spatial* | Attrib: Added *training_mode*, changed the name of the type constraint (Not the type itself) for mean and var. Changed the name of output variables in training| Changed the name of the type constraint (Not the type itself)for mean and var, and also for scale and B(bias) |

Also, I have a question.
Should I leave logic in the code splitting for training and Inference, even since the training is not implemented, and would lead the code to an assertion?
like:

```cpp
ov::OutputVector batch_norm(const ov::frontend::onnx::Node& node) {
    ov::OutputVector inputs{node.get_ov_inputs()};
    auto x = inputs.at(0);
    auto scale = inputs.at(1);
    auto bias = inputs.at(2);
    auto mean = inputs.at(3);
    auto var = inputs.at(4);

    double epsilon{node.get_attribute_value<double>("epsilon", 1e-5)};
    double momentum{node.get_attribute_value<double>("momentum", 0.9)};
    double training_mode{node.get_attribute_value<double>("training_mode", 0)}; // 0 = Training
    if (training_mode == true) {
        CHECK_VALID_NODE(node, node.get_outputs_size() == 1, "Training mode of BatchNormalization is not supported.");
        // return {std::make_shared<v5::BatchNormTraining>(x, scale, bias, mean, var, epsilon, momentum)}; // not implemented
    } else {
        CHECK_VALID_NODE(node, node.get_outputs_size() == 1, "Inference set in flag training_mode, but expected more than a single output");
        return {std::make_shared<v5::BatchNormInference>(x, scale, bias, mean, var, epsilon)};
    }
```
Or something without that much logic, like:
```cpp
ov::OutputVector batch_norm(const ov::frontend::onnx::Node& node) {
    ov::OutputVector inputs{node.get_ov_inputs()};
    auto x = inputs.at(0);
    auto scale = inputs.at(1);
    auto bias = inputs.at(2);
    auto mean = inputs.at(3);
    auto var = inputs.at(4);

    double epsilon{node.get_attribute_value<double>("epsilon", 1e-5)};
    // Attribute "spatial" is ignored, as we only support inference mode of
    // BatchNormalization

    CHECK_VALID_NODE(node, node.get_outputs_size() == 1, "Training mode of BatchNormalization is not supported.");

    return {std::make_shared<v5::BatchNormInference>(x, scale, bias, mean, var, epsilon)};
}
```

**Comment by gkrivor** [comment]
Hi @Bepitic ,

I suggest to extended check in second option:
CHECK_VALID_NODE(node, training_mode == false && node.get_outputs_size() == 1, "Training mode of BatchNormalization is not supported.");
